File: src/main/java/com/littlegit/server/authfilter/AuthFilter.java
package com.littlegit.server.authfilter;

import com.littlegit.server.model.user.AuthRole;
import com.littlegit.server.service.AuthService;

import javax.annotation.Priority;
import javax.inject.Inject;
import javax.ws.rs.Priorities;
import javax.ws.rs.container.ContainerRequestContext;
import javax.ws.rs.container.ContainerRequestFilter;
import javax.ws.rs.container.ResourceInfo;
import javax.ws.rs.core.Context;
import java.io.IOException;
import java.lang.reflect.AnnotatedElement;
import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

@Secured
@Priority(Priorities.AUTHORIZATION)
public class AuthFilter implements ContainerRequestFilter {

    @Context
    private ResourceInfo resourceInfo;
    private AuthService authService;

    @Inject
    public AuthFilter(AuthService authService) {
        this.authService = authService;
    }

    @Override
    public void filter(ContainerRequestContext requestContext) throws IOException {
        List<AuthRole> allowedRoles = getAllowedRoles();
        checkPermissions(allowedRoles);
    }

    private List<AuthRole> getAllowedRoles() {
        // Get the resource class which matches with the requested URL
        // Extract the roles declared by it
        Class<?> resourceClass = resourceInfo.getResourceClass();
        List<AuthRole> classRoles = extractRoles(resourceClass);

        // Get the resource method which matches with the requested URL
        // Extract the roles declared by it
        Method resourceMethod = resourceInfo.getResourceMethod();
        List<AuthRole> methodRoles = extractRoles(resourceMethod);

        // Check if the user is allowed to execute the method
        // The method annotations override the class annotations
        if (methodRoles.isEmpty()) {
            return classRoles;
        } else {
            return methodRoles;
        }
    }
    // Extract the roles from the annotated element
    private List<AuthRole> extractRoles(AnnotatedElement annotatedElement) {
        if (annotatedElement == null) {
            return new ArrayList<>();
        } else {
            Secured secured = annotatedElement.getAnnotation(Secured.class);
            if (secured == null) {
                return new ArrayList<>();
            } else {
                AuthRole[] allowedRoles = secured.value();
                return Arrays.asList(allowedRoles);
            }
        }
    }

    private void checkPermissions(List<AuthRole> allowedRoles) {
        System.out.println(allowedRoles);
    }
}

File: src/main/java/com/littlegit/server/authfilter/AuthFilter.kt
package com.littlegit.server.authfilter

import com.littlegit.server.application.exception.UserForbiddenException
import com.littlegit.server.application.exception.UserUnauthorizedException
import com.littlegit.server.model.user.AuthRole
import com.littlegit.server.service.AuthService
import java.io.IOException
import java.lang.reflect.AnnotatedElement
import java.security.Principal
import java.util.*
import javax.annotation.Priority
import javax.inject.Inject
import javax.ws.rs.Priorities
import javax.ws.rs.container.ContainerRequestContext
import javax.ws.rs.container.ContainerRequestFilter
import javax.ws.rs.container.ResourceInfo
import javax.ws.rs.core.Context
import javax.ws.rs.core.HttpHeaders
import javax.ws.rs.core.SecurityContext


@Secured
@Priority(Priorities.AUTHORIZATION)
open class AuthFilter @Inject
constructor(private val authService: AuthService) : ContainerRequestFilter {

    @Context
    private lateinit var resourceInfo: ResourceInfo

    private val allowedRoles: List<AuthRole>
        get() {

            val resourceClass = resourceInfo.resourceClass
            val classRoles = extractRoles(resourceClass)

            val resourceMethod = resourceInfo.resourceMethod
            val methodRoles = extractRoles(resourceMethod)
            return if (methodRoles.isEmpty()) {
                classRoles
            } else {
                methodRoles
            }
        }

    @Throws(IOException::class)
    override fun filter(requestContext: ContainerRequestContext) {
        val authHeader = requestContext.getHeaderString(HttpHeaders.AUTHORIZATION)

        if (authHeader?.startsWith(AuthConstants.AuthScheme) == true) {

            val token = authHeader.removePrefix("${AuthConstants.AuthScheme} ")
            val user = authService.getUserForToken(token)

            if (!user.hasAnyRoleOf(allowedRoles)) {
                throw UserForbiddenException()
            }

            requestContext.securityContext = object: SecurityContext {
                override fun isUserInRole(p0: String?) = true
                override fun getAuthenticationScheme(): String = AuthConstants.AuthScheme
                override fun getUserPrincipal(): Principal = user
                override fun isSecure(): Boolean = requestContext.securityContext.isSecure
            }
        } else {
            throw UserUnauthorizedException()
        }
    }

    // Extract the roles from the annotated element
    private fun extractRoles(annotatedElement: AnnotatedElement?): List<AuthRole> {
        return if (annotatedElement == null) {
            ArrayList()
        } else {
            val secured = annotatedElement.getAnnotation(Secured::class.java)
            if (secured == null) {
                ArrayList()
            } else {

                val allowedRoles = secured.value
                Arrays.asList(*allowedRoles)
            }
        }
    }
}